pkg/controller/expcertificates/issuing/secret.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2020 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This file defines methods used for PKCS#12 support.
+// This is an experimental feature and the contents of this file are intended
+// to be absorbed into a more fully fledged implementing ahead of the v0.15
+// release.
+// This should hopefully not exist by the next time you come to read this :)
+package issuing
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apiutil "github.com/jetstack/cert-manager/pkg/api/util"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha2"
+	cmmeta "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	utilpki "github.com/jetstack/cert-manager/pkg/util/pki"
+)
+
+// updateSecretData will ensure the Secret resource contains the given secret
+// data as well as appropriate metadata.
+// If the Secret resource does not exist, it will be created.
+// Otherwise, the existing resource will be updated.
+// The first return argument will be true if the resource was updated/created
+// without error.
+// updateSecretData will also update deprecated annotations if they exist.
+func (c *controller) updateSecretData(ctx context.Context, namespace string, crt *cmapi.Certificate, data secretData) error {
+	// Fetch a copy of the existing Secret resource
+	secret, err := c.secretLister.Secrets(crt.Namespace).Get(crt.Spec.SecretName)
+	if !apierrors.IsNotFound(err) && err != nil {
+		// If secret doesn't exist yet, then don't error
+		return err
+	}
+
+	secretExists := true
+
+	// If the seret does not exist yet, then we need to create one
+	if secret == nil {
+		secret = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      crt.Spec.SecretName,
+				Namespace: namespace,
+			},
+			Type: corev1.SecretTypeTLS,
+		}
+
+		secretExists = false
+	}
+
+	// secret will be overwritten by 'existingSecret' if existingSecret is non-nil
+	if c.enableSecretOwnerReferences {
+		secret.OwnerReferences = []metav1.OwnerReference{*metav1.NewControllerRef(crt, certificateGvk)}
+	}
+
+	newSecret := secret.DeepCopy()
+	err = c.setSecretValues(crt, newSecret, data)
+	if err != nil {
+		return err
+	}
+	if reflect.DeepEqual(secret, newSecret) {
```

Comment:
This behaviour does make sense, although I wonder if it will 'hide' real errors in our code/state machine.. perhaps one to try and tackle later, but for now maybe add a comment and possibly a log message in case this case occurs? (maybe at DebugLevel..)

Reply:
Hm, this check also won't work when P12/JKS support is enabled, as these formats use a random parameter when encrypting the keystore meaning the secretes will never be *equal* despite containing equal private keys/certs.

Given that, and given we don't have exhaustive testing around P12/JKS right now to identify how serious of an issue this is, an implementation that doesn't rely on this check would be best if possible!